runtime/runtime-params-estimator/src/trie.rs

```diff
@@ -0,0 +1,413 @@
+                num_warmup_values,
+                data_spread_factor,
+                spoil_l3,
+            );
+            let p_results = percentiles(results, &[0.5, 0.9, 0.99, 0.999]);
+            if debug {
+                print!("{:<32}", "Best Case (from memory)");
+                for cost in p_results.iter() {
+                    print!("{:>8} ", cost.to_gas() / 1_000_000);
+                }
+                println!();
+            }
+        }
+        // Best-case
+        {
+            let spoil_l3 = false;
+            let num_warmup_values = 1;
+            let num_values = 128;
+            let iters = 30; // For estimation speed only, should not affect results
+            let data_spread_factor = 1;
+            let results = read_node_from_chunk_cache_ext(
+                testbed,
+                iters,
+                num_values,
+                num_warmup_values,
+                data_spread_factor,
+                spoil_l3,
+            );
+            let p_results = percentiles(results, &[0.5, 0.9, 0.99, 0.999]);
+            if debug {
+                print!("{:<32}", "Best Case");
+                for cost in p_results.iter() {
+                    print!("{:>8} ", cost.to_gas() / 1_000_000);
+                }
+                println!();
+            }
+        }
+    }
+
+    base_case
+}
+
+fn read_node_from_chunk_cache_ext(
+    testbed: &mut Testbed,
+    iters: usize,
+    // How many values are read after each other. The higher the number, the
+    // larger the amortization effect is.
+    num_values: usize,
+    // Values to read before measurement, to warm up CPU caches with data
+    // structures used in looking up trie nodes
+    num_warmup_values: usize,
+    // Spread factor to reduce data locality
+    data_spread_factor: usize,
+    // Before measuring, completely overwrite L3 CPU cache content
+    spoil_l3: bool,
+) -> Vec<GasCost> {
+    // Trie nodes are largest when they hold part of a storage key (Extension or
+    // Leaf) and keys can be up to 2kiB. Therefore, this is about the maximum
+    // size possible.
+    let value_len: usize = 2048;
+
+    // Prepare a data buffer that we can read again later to overwrite CPU caches
+    let assumed_max_l3_size = 128 * 1024 * 1024;
+    let dummy_data = testbed.transaction_builder().random_vec(assumed_max_l3_size);
+
+    (0..iters)
+        .map(|i| {
+            let tb = testbed.transaction_builder();
+            let signer = tb.random_account();
+            let values_inserted = num_values * data_spread_factor;
+            let values: Vec<_> = (0..values_inserted)
+                .map(|_| {
+                    let v = tb.random_vec(value_len);
+                    let h = hash(&v);
+                    let node = RawTrieNode::Extension(v, h);
+                    let node_with_size = RawTrieNodeWithSize { node, memory_usage: 1 };
+                    node_with_size.encode().unwrap()
+                })
```

Comment:
Wait, what is happening here? I am hopelessly confused here -- to me, it seems that this block should be roughly equivalent to just `tb.random_vec(value_len)`?

Like, we encode a raw trie node, but then use this thing as a *value* we just insert? Ie, what we get in a database is a trie node whose payload is an encoded trie node, no?

Reply:
Oof, looking back at this, I am confused myself. :sweat_smile: At the very least, it needs some comments.

This evolved as @Longarithm and I were trying different estimation strategies. I actually didn't write the difficult part of the code but I hope understood it right. Let me explain.

Yes, it does store an encoded trie node as the value of a node. And it is confusing, yet intentional. 

The goal here is two-fold.
1. Use the trie cache directly (i.e. without the trie) but keep the encoding and decoding logic. Basically, emulate the normal behaviour to lookup a trie node, which is this:
```rust
fn retrieve_node(&self, hash: &CryptoHash) -> Result<TrieNodeWithSize, StorageError> {
    if *hash == Trie::empty_root() {
        return Ok(TrieNodeWithSize::empty());
    }
    let bytes = self.storage.retrieve_raw_bytes(hash)?;
    match RawTrieNodeWithSize::decode(&bytes) {
        Ok(value) => Ok(TrieNodeWithSize::from_raw(value)),
        Err(_) => Err(StorageError::StorageInconsistentState(format!(
            "Failed to decode node {}",
            hash
        ))),
    }
}
```
2. Read a value that represents the worst-case for cached node lookup. And the worst-case is, an extension trie node with a 2kB long key in it. Why? Because reading most nodes is dominated by other costs, such as `storage_read_value_byte` for leaf nodes and `storage_read_key_byte` for branches and extension we just iterate over. However, when we read an extension node form the cache because we look up the value stored _at_ the extension point, in that case we read the key extension of the value _below_ the node as well. And this key extension can be potentially 2k bytes long and is not part of the gas equation. So that is the worst-case we care about in this estimation.

Generating many different extension nodes inside the trie is a bit tricky, since they interact with each other and we don't want keys to become shorter than 2kB. So it would be nice to have a more direct way to just insert trie nodes that look like our worst case, so we read those into the cache.

The tricky part on the implementation side is, the interface `trait TrieStorage` only reads values. Encoding and storing of values is done on layers above that. Thus, the trick is to insert an encoded trie node as a value through normal block processing. When inserting `("key", "value")` like that, a bunch of actual trie nodes for `"key"` are inserted to RocksDB, all of which we ignore completely. All we need is the raw `"value"`, which is also inserted by its hash as the key.

Now, since we know the hash, we can obtain our raw node value through the trie storage interface and decode it like a real trie node. 

I admit, this is rather insane... But the thing we want to measure it too small that we could allow extra overhead. Any kind of subtraction between two costs we tried was highly unstable.

Reply:
That's smart! But would definitely benefit from a comment :-)